adaptdl/adaptdl/torch/data.py

```diff
@@ -405,11 +451,20 @@ def __iter__(self):
             while not done:
                 self.sampler.set_epoch(epoch, index=self._elastic.current_index)  # noqa: E501
                 self.batch_sampler.batch_size = self._elastic._sync_local_bsz()
+                base_iterations = (len(self.sampler)
+                                   / self.batch_sampler.batch_size)
+                iterations = int(
+                    base_iterations -
+                    base_iterations % self._elastic._grad_acc_steps)
                 for idx, batch in enumerate(super().__iter__()):
+                    if idx > iterations:
+                        break
                     with self._elastic.profile():
                         yield batch
                         # Increment by the number of data samples processed
-                        self._elastic.current_index += self.current_batch_size  # noqa: E501
+                        self._elastic.current_index += (
+                            self._elastic._current_local_bsz
+                            * adaptdl.env.num_replicas())
```

Comment:
I think we should try hard to avoid modifications to `AdaptiveDataLoader`, since that means we are increasing the complexity of implementing custom data loaders. Conceptually, the `AdaptiveDataLoader` can be agnostic to gradient accumulation, and simply produce local batch of the right size, as specified by the `AdaptiveDataLoaderHelper`.